libbeat/processors/add_cloud_metadata/provider_azure_vm.go

```diff
@@ -50,11 +84,138 @@ var azureVMMetadataFetcher = provider{
+		fetcher, err := newAzureMetadataFetcher("azure", hfetcher)
+		if err != nil {
+			return fetcher, err
+		}
+		gfetcher, err := newGenericMetadataFetcher(config, "azure", azGenSchema, fetcher.fetchAzureClusterMeta)
+		if err != nil {
+			return fetcher, err
+		}
+		fetcher.genericMetadataFetcher = gfetcher
 		return fetcher, err
 	},
 }
+
+func (az *azureMetadataFetcher) fetchMetadata(ctx context.Context, client http.Client) result {
+	res := result{provider: az.provider, metadata: mapstr.M{}, err: nil}
+	logger := logp.NewLogger("add_cloud_metadata")
+	httpRes := az.httpMetadataFetcher.fetchMetadata(ctx, client)
+	if httpRes.err != nil {
+		res.err = httpRes.err
+		return res
+	}
+	res.metadata = httpRes.metadata
+	az.httpMeta = httpRes.metadata
+	gRes := az.genericMetadataFetcher.fetchMetadata(ctx, client)
+	if gRes.err != nil {
+		logger.Warnf("Failed to get additional AKS Cluster meta: %+v", gRes.err)
+		return res
+	}
+
+	res.metadata.DeepUpdate(gRes.metadata)
+	return res
+}
+
+// getAzureCredentials returns credentials to connect to Azure
+// env vars TENANT_ID, CLIENT_ID and CLIENT_SECRET are required
+// if not set, NewDefaultAzureCredential method will be used
+func getAzureCredentials(logger *logp.Logger) (azcore.TokenCredential, error) {
+	if os.Getenv("TENANT_ID") != "" && os.Getenv("CLIENT_ID") != "" && os.Getenv("CLIENT_SECRET") != "" {
+		return azidentity.NewClientSecretCredential(os.Getenv("TENANT_ID"), os.Getenv("CLIENT_ID"), os.Getenv("CLIENT_SECRET"), nil)
+	} else {
+		logger.Debugf("No Client or Tenant configuration provided. Retrieving default Azure credentials")
+		return azidentity.NewDefaultAzureCredential(nil)
+	}
+}
+
+// getAKSClusterNameId returns the AKS cluster name and Id for a given resourceGroup
+func getAKSClusterNameId(ctx context.Context, logger *logp.Logger, clusterClient *armcontainerservice.ManagedClustersClient, resourceGroupName string) (string, string, error) {
+	pager := clusterClient.NewListPager(nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return "", "", fmt.Errorf("failed to advance page: %w", err)
+		}
+		for _, v := range page.Value {
+			if *v.Properties.NodeResourceGroup == resourceGroupName {
+				return *v.Name, *v.ID, nil
+			}
+
+		}
+	}
+	return "", "", nil
+}
+
+// fetchAzureClusterMeta fetches metadata of Azure Managed Clusters using azure sdk.
+func (az *azureMetadataFetcher) fetchAzureClusterMeta(
+	ctx context.Context,
+	client http.Client,
+	result *result,
+) {
+	logger := logp.NewLogger("add_cloud_metadata")
+	subscriptionId, _ := az.httpMeta.GetValue("cloud.account.id")
+	resourceGroupName, _ := az.httpMeta.GetValue("cloud.resourceGroup.name")
+	strResourceGroupName := ""
+	if val, ok := resourceGroupName.(string); ok {
+		strResourceGroupName = val
+	}
+	strSubscriptionId := ""
+	if val, ok := subscriptionId.(string); ok {
+		strSubscriptionId = val
+	}
+	// if subscriptionId cannot be retrieved from metadata endpoint, try with env var
```

Comment:
in what cases `subscriptionId` might not be retrieved from metadata? 
we are also not mentioning anywhere that `SUBSCRIPTION_ID` can/should be provided as env var

Reply:
As far as I have seen, in no way. I just wanted to give an extra option. The user may check the logs and find that they have an alternative. I am not sure where to document this without getting in so many details of the implementation

Reply:
> I am not sure where to document this without getting in so many details of the implementation

I see
I still think that an extra option might be unnecessary and add more confusing - providing hidden alternative, that solves unknown issue

but it is not a blocker for this PR

Reply:
I will remove the env var option as it may be confusing.